src/Catalyst.Core.Lib/Dag/DagNode.cs

```diff
@@ -0,0 +1,371 @@
+    /// <remarks>
+    ///     A <b>DagNode</b> has opaque <see cref="DagNode.DataBytes" />
+    ///     and a set of navigable <see cref="DagNode.Links" />.
+    /// </remarks>
+    [DataContract]
+    public sealed class DagNode : IDagNode, IEquatable<DagNode>
+    {
+        private Cid _id;
+        private string _hashAlgorithm = MultiHash.DefaultAlgorithmName;
+        private long? _size;
+
+        /// <summary>
+        ///     Create a new instance of a <see cref="DagNode" /> with the specified
+        ///     <see cref="DagNode.DataBytes" /> and <see cref="DagNode.Links" />
+        /// </summary>
+        /// <param name="data">
+        ///     The opaque data, can be <b>null</b>.
+        /// </param>
+        /// <param name="links">
+        ///     The links to other nodes.
+        /// </param>
+        /// <param name="hashAlgorithm">
+        ///     The name of the hashing algorithm to use; defaults to
+        ///     <see cref="MultiHash.DefaultAlgorithmName" />.
+        /// </param>
+        public DagNode(byte[] data,
+            IEnumerable<IMerkleLink> links = null,
+            string hashAlgorithm = MultiHash.DefaultAlgorithmName)
+        {
+            DataBytes = data ?? new byte[0];
+            Links = (links ?? new DagLink[0])
+               .OrderBy(link => link.Name ?? "");
+            _hashAlgorithm = hashAlgorithm;
+        }
+
+        /// <summary>
+        ///     Creates a new instance of the <see cref="DagNode" /> class from the
+        ///     specified <see cref="Stream" />.
+        /// </summary>
+        /// <param name="stream">
+        ///     A <see cref="Stream" /> containing the binary representation of the
+        ///     <b>DagNode</b>.
+        /// </param>
+        public DagNode(Stream stream) { Read(stream); }
+
+        /// <summary>
+        ///     Creates a new instance of the <see cref="DagNode" /> class from the
+        ///     specified <see cref="CodedInputStream" />.
+        /// </summary>
+        /// <param name="stream">
+        ///     (
+        ///     A <see cref="CodedInputStream" /> containing the binary representation of the
+        ///     <b>DagNode</b>.
+        /// </param>
+        public DagNode(CodedInputStream stream) { Read(stream); }
+
+        /// <inheritdoc />
+        [DataMember]
+        public IEnumerable<IMerkleLink> Links { get; private set; }
+
+        /// <inheritdoc />
+        [DataMember]
+        public byte[] DataBytes { get; private set; }
+
+        /// <inheritdoc />
+        public Stream DataStream => new MemoryStream(DataBytes, false);
+
+        /// <summary>
+        ///     The serialised size in bytes of the node.
+        /// </summary>
+        [DataMember]
+        public long Size
+        {
+            get
+            {
+                if (!_size.HasValue)
+                {
+                    ComputeSize();
+                }
+
+                return _size ?? throw new ArgumentNullException();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use a constructor overloads that allows a more meaningful exception message to be provided.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)